src/transformers/models/llava/modeling_llava.py

```diff
@@ -409,25 +477,79 @@ def forward(
         )
 
         if inputs_embeds is None:
-            # 1. Extra the input embeddings
+            # 1. Extract the input embeddings
             inputs_embeds = self.get_input_embeddings()(input_ids)
 
             # 2. Merge text and images
             if pixel_values is not None and input_ids.shape[1] != 1:
-                image_outputs = self.vision_tower(pixel_values, output_hidden_states=True)
-                # this is not memory efficient at all (output_hidden_states=True) will save all the hidden stated.
-                selected_image_feature = image_outputs.hidden_states[vision_feature_layer]
-
-                if vision_feature_select_strategy == "default":
-                    selected_image_feature = selected_image_feature[:, 1:]
-                elif vision_feature_select_strategy == "full":
-                    selected_image_feature = selected_image_feature
+                if pixel_values.ndim == 5:
+                    concat_images = torch.cat(list(pixel_values), dim=0)
+                    image_features = self.vision_tower(concat_images, output_hidden_states=True)
+
+                    selected_image_feature = image_features.hidden_states[vision_feature_layer]
+
+                    if vision_feature_select_strategy == "default":
+                        selected_image_feature = selected_image_feature[:, 1:]
+                    elif vision_feature_select_strategy == "full":
+                        selected_image_feature = selected_image_feature
+                    else:
+                        raise ValueError(
+                            f"Unexpected select feature strategy: {self.config.vision_feature_select_strategy}"
+                        )
+                    image_features = self.multi_modal_projector(selected_image_feature)
+
+                    split_sizes = [image.shape[0] for image in pixel_values]
+                    image_features = torch.split(image_features, split_sizes, dim=0)
+
+                    # NOTE we only support multimodal_patch_merge_type == "spatial_unpad"
+                    new_image_features = []
+                    for image_idx, image_feature in enumerate(image_features):
```

Comment:
This is going to be super slow iterating over the image like this. Looking at the code, it looks like this can all be vectorized

Reply:
I'm afraid this can't be vectorized, since you need to go over each of the individual image features since the image sizes differ per image in the batch. This was taken from the original implementation here: https://github.com/haotian-liu/LLaVA/blob/5d8f1760c08b7dfba3ae97b71cbd4c6f17d12dbd/llava/model/llava_arch.py#L166

Reply:
the goal is to not follow the original implementation. Which means more work for use of course but we did not follow the original implementation either and successfully vectorized the forward pass 😉

Reply:
Could both of you elaborate a bit on how this code can be vectorized? Because the `images_sizes` can be different for each image in the batch, so I'm not seeing how this can be easily vectorized

Reply:
Basically this line differs per image in the batch, preventing batching: https://github.com/NielsRogge/transformers/blob/add_llava_1_6/src/transformers/models/llava/modeling_llava.py#L508

Reply:
I have no idea what this is doing so I don't know, but maybe we can change the paradigm in a way that prevents us from having to do this. 
Think it that way: if the processor has the number of images, the size of the images and the raw text, it can act on a text level, which is super super fast even if you do it sequentially. 
If the pre-processing is just giving more embedding tokens to images (so a bigger image has 512 + more tokens, while a smaller image has 512 - less tokens) then we can just do that in the processor. Replace the `<image>` token with `num_image_tokens[i] * "<fake_image_token"` which should be alot lot simpler

Reply:
@NielsRogge @ArthurZucker Hey, nice works here. I'm re-implementing your model code so that that this unpadding steps will be more efficient.
It's true that you can't vectorize the images completely, but you can do so by pre-calculating the indices and the final size of the full_image_features (for all images). Here is the steps:

- From the image_sizes, use your for loop to calculate how many tokens each image has, how many patches (it's not always 5) each image has, and the slice indices of each patche (incase of unpadding). This can be done with cpu because you don't deal with tensor here.
- Create a single `full_image_embeds` of size (all_img_tokens, embed_dim).
- for loop through image features and copy the indexed img features to the correct position of `full_image_embeds`. (*This will be faster than your implementation with `torch.cat` as it will create new memory each time*, and this copying can be done asynchronously by GPU)
- finally, with the indices, copy `full_image_embeds` to `final_embeddings` where <image> is.

Reply:
Awesome - thanks for sharing @nxphi47! Do you have any numbers on how much this speeds things up?

Reply:
Hi @nxphi47 if you're willing to open a PR regarding this - that'd be great! I think this could also solve batched generation which is requested at #29832

Reply:
@NielsRogge Working on it. My implementation is based on your previously cancelled PR, so I'm adapting it to the new llava-next. 

As for #29832 , the idea is concatenating pixel_values along patch dim (e.g `(8, 3, 336, 336)` for 2 images with patch length 5 and 3), and not stacking into a batch dimension.

Reply:
Ok so based on offline discussion with @ArthurZucker, the idea is that this could already be handled in the tokenizer (`LlavaNextProcessor`) without the model requiring the `image_sizes` argument. Basically there we can already calculate the amount of image tokens per image in a batch, and add them to the `input_ids`. 

The model will then replace the special image tokens in the `input_ids` by the image features.

Is this similar to what you were thinking of?

Reply:
@NielsRogge cannot do that without image_sizes, because you need to inject the newline_embedding between patches and unpad the patch vision feature after vision_tower.

Reply:
What if you: 
1. Process the images and get the `image_sizes`
2. Process the strings, replacing fake tokens with `image_sizes[batch][image]` image tokens + `new_line`
then the new_line embedding will be correctly placed no?

Reply:
Yeah I was also think that if you add the `new_line` token to the `input_ids`, then in the model this could be replaced by the `image_newline` trainable parameter

Reply:
@NielsRogge @ArthurZucker  You still have to handle unpadding of image features.

Anyway, I port my implementation here: https://github.com/huggingface/transformers/pull/29850

I have not done testing and bugs are expected, but this should essentially tell you how I'm doing it.

It supports:
- Batched forward with multiple image of different sizes (number of patches).
- Training, for cases without any image.
- multi-image in same sequence. e.g: `[" <image> <image> the first image is a dog while the second is a cat", "    <image> <image> <image> <image these 4 image are..."]`